src/GitHub/App.hs

```diff
@@ -43,95 +41,75 @@ type BaseURL = Text
+-- and also returns that new token. 
 authenticateInstallation :: InstallationAuth -> IO Auth
 authenticateInstallation instAuth = do 
-    gtaToken    <- readIORef (instAuth ^. token)
-    currentTime <- getPOSIXTime
-    if (expireAt gtaToken) - currentTime >= bufferTime 
-    then return $ OAuth $ encodeUtf8 $ instToken gtaToken
+    gtaToken    <- readMVar (instAuth ^. token)
+    currentTime <- getCurrentTime
+    if getExpirationTime gtaToken `diffUTCTime` currentTime >= bufferTime 
+    then return $ OAuth $ getToken gtaToken
     else do 
-        let applicationId = instAuth ^. appId
-        OAuth <$> encodeUtf8 <$> instToken <$> (join $ readIORef <$> renewInstAuthToken instAuth) 
+        updatedToken   <- takeMVar (instAuth ^. token)
+        renewInstAuthToken instAuth
+        updatedToken   <- readMVar (instAuth ^. token)
+        return $ OAuth $ getToken updatedToken
 
+-- | Creates new InstallationAuth value. Useing this function is the only way to create 
+-- InstallationAuth value, because we don't export constructors. 
+createInstAuth :: Int -> PrivateKey -> InstallationId -> IO InstallationAuth 
+createInstAuth appId key instId = do 
+    token  <- newEmptyMVar
+    let instAuth = InstallationAuth appId key instId token
+    _      <- renewInstAuthToken instAuth
+    return instAuth
 
--- | Gets new token from github and writes it to IORef in given InstallationAuth. Also returns that token. 
-renewInstAuthToken :: InstallationAuth -> IO (IORef GTAToken) 
+-- | Gets new token from github and writes it to IORef in given InstallationAuth. 
+renewInstAuthToken :: InstallationAuth -> IO () 
 renewInstAuthToken instAuth = do 
-    time    <- getPOSIXTime
+    time    <- getCurrentTime
     let tkn = instAuth ^. token
-    case (makeJWT time (instAuth ^. appId) (instAuth ^. appPrivateKey)) of 
-        Left  err  -> throwM (FailedGenerateJWT err) 
-        Right jwt  -> do 
-            t <- request (https baseURL /: "installations" /: (instAuth ^. installationId) /: "access_tokens") mempty jwt    
-            writeIORef tkn t 
-            return tkn
-
--- | Creates new InstallationAuth value. Useing this function is the only way to create InstallationAuth value, 
--- because we don't export constructors. 
-createInstAuth :: Int -> PrivateKey -> InstallationId -> IO InstallationAuth 
-createInstAuth appId key instId = do 
-    time    <- getPOSIXTime
-    case (makeJWT time appId key) of 
-        Left  err  -> throwM (FailedGenerateJWT err) 
-        Right jwt  -> do 
-            t      <- request (https baseURL /: "installations" /: instId /: "access_tokens") mempty jwt
-            token  <- newIORef t
-            return $ InstallationAuth appId key instId token
+    let jwt = makeJWT time (instAuth ^. appId) (instAuth ^. appPrivateKey) 
+    t <- request (https baseURL /: "installations" /: (instAuth ^. installationId) /: "access_tokens") mempty jwt    
+    putMVar tkn t 
```

Comment:
Ok. I did it because i always remove value from MVar before calling this function and it can not be used outside of this module. That is why i did it in this way.

Reply:
Yes, that’s why I didn’t say it was incorrect, I only said it was dangerous and should have been clearly documented ;).